File: solutions/0042-trapping-rain-water/solution.cpp
class Solution {
public:
    int trap(vector<int>& height) {
         int n = height.size();
        int sum = 0;
        int lm[n];
        int rm[n];
        int max = 0;
        for(int i = 0;i<n;i++){
           if(height[i]  > max){
               max =  height[i];
           }
           lm[i] = max;
        }
          max = 0;
          for(int i = n-1;i>=0;i--){
           if(height[i]  > max){
               max =  height[i];
           }
            rm[i] = max;
        }
        for (int i = 1; i < height.size()-1; i++){
        int wt = min(lm[i],rm[i]);
        if(wt-height[i] > 0){
         sum += wt-height[i];
            
        }
        }

        return sum;
    }
};

File: solutions/0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] height) {
        int n = height.length - 1;
        int[] lmap = new int[n+1];
        int[] rmap = new int[n+1];
        int lmax = 0;
        int rmax = 0;
        int water=  0;
        for(int i = 0;i<=n;i++){
            int capacity = lmax-height[i];
            if(capacity<0){
                capacity = 0;
            }
            lmap[i] = capacity;
            lmax = Math.max(lmax,height[i]);
            
            capacity = rmax-height[n-i];
             if(capacity<0){
                capacity = 0;
            }
            rmap[n-i] = capacity;
            rmax = Math.max(rmax,height[n-i]);
        }
        for(int i = 0;i<n;i++){
            water += Math.min(lmap[i],rmap[i]);
        }
        return water;
    }
}
